thetis/timeintegrator.py

```diff
@@ -68,6 +72,27 @@ def __init__(self, equation, solution, fields, dt, solver_parameters={}):
         self.solver_parameters = {}
         self.solver_parameters.update(solver_parameters)
 
+        if self.equation.__class__.__name__ == "ShallowWaterEquations":
+            self.momentum_residual = ShallowWaterMomentumResidual(
```

Comment:
The problem here is that the time integrators are aware of the equation being solved. It becomes difficult to write/maintain such if clauses for all the equations and time integrators.
Currently the time integrator classes take an `Equation` as an argument in the constructor. You could add (an optional?) residual argument as well. The correct residual object can then be created simultaneously with the equation (this happens in the solver class).